gswrap/__init__.py

```diff
@@ -804,6 +798,9 @@ def _download(self,
                 if no_clobber and file_path.exists():
                     continue
 
+                file_path_parent = pathlib.Path(file_path.parent)
```

Comment:
Here you can have a similar structure as above:
* prepare the parameters
* prepare the setting (by creating the directories)
* execute with multiple threads

You can even mark the sections with 

```
##
# Collect sources/destinations
##

##
# Make parent directories
##

##
# Execute
##
```